### PR TITLE
BLD Extract out package build configs from Makefile.envs

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

Related: #4557 

This PR extracts variables that are not used in Pyodide main module, but only in pyodide-build from Makefile.envs. This change allows us to manage changes to pyodide-build independently of the Makefile.

Inside pyodide-build, I made a new module `ConfigManager`, and made all variables used in pyodide-build to be managed by that module. There is much room for improvement in this module, but I am opening this PR to minimize the change in a single PR. Once this PR is merged, I'll make more changes as a follow-up.

### Checklists

- [x] Add / update tests

### CODE REVIEW

File: pyodide-build/pyodide_build/config.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+            BUILD_VAR_TO_KEY[key]: env[key] for key in env if key in BUILD_VAR_TO_KEY
+        }
+
+    def _load_config_file(self) -> dict[str, str]:
+        # TODO: Implement this
+        return {}
+
+    @property
+    def config(self) -> dict[str, str]:
+        return self._config.copy()
```
Comment (hoodmane) [inline pyodide-build/pyodide_build/config.py:96]: It isn't really ideal that this copies, for example in `to_env` we make two copies of the dictionary, once to access `self.config`. What about 
```py
from types import MappingProxyType
# ...
    @property
    def config(self) -> dict[str, str]:
        return MappingProxyType(self._config)
```

**Review by hoodmane — APPROVED** [review]
Thanks!

File: pyodide-build/pyodide_build/build_env.py [hunk 2]
```diff
@@ -166,83 +125,31 @@ def in_xbuildenv() -> bool:
...
-    env["PYTHONPATH"] = ":".join(pythonpath)
+    config_manager = ConfigManager(pyodide_root)
+    env = config_manager.to_env()
+
+    env |= {
+        # This environment variable is used for packages to detect if they are built
+        # for pyodide during build time
+        "PYODIDE": "1",
+        # This is the legacy environment variable used for the aforementioned purpose
+        "PYODIDE_PACKAGE_ABI": "1",
```
[the remark this answers is not in the document]
Comment (ryanking13) [reply pyodide-build/pyodide_build/build_env.py:140]: Yes, let's remove it in the next couple of releases. BTW, I found it is still used in [scikit-learn](https://github.com/scikit-learn/scikit-learn/blob/79e14a77565e28fc716cd292a80f35a92efa1ca4/sklearn/_build_utils/pre_build_helpers.py#L63).

File: pyodide-build/pyodide_build/config.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+
+class ConfigManager:
+    """
+    Configuration manager for Package build process.
+
+    The configuration manager is responsible for loading configuration from various sources.
+    The configuration can be loaded from the following sources (in order of precedence):
+
+        1. Command line arguments
```
[the remark this answers is not in the document]
Comment (ryanking13) [reply pyodide-build/pyodide_build/config.py:16]: Oh, yeah. I think cmdline args should be handled here, too, for consistency. It is TODO and not included in this PR. Let me add a comment about it.

File: pyodide-build/pyodide_build/config.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    """
+
+    def __init__(self, pyodide_root: Path):
+        self.pyodide_root = pyodide_root
+        self._config = {
+            **self._load_default_config(),
+            **self._load_makefile_envs(),
+            **self._load_config_file(),
+            **self._load_config_from_env(dict(os.environ)),
+        }
```
[the remark this answers is not in the document]
Comment (ryanking13) [reply pyodide-build/pyodide_build/config.py:30]: I was too lazy to find the type for `os.environ` 😅 . Will change it to use os.environ directly without copying it.

**Pull request merged**